crates/wasi/src/preview2/pipe.rs

```diff
@@ -7,227 +7,835 @@
-///
-/// This reads from a source that implements the [`Read`] trait. It
-/// also requires the [`ReadReady`] trait, which is implemented for many
-/// popular `Read`-implementing types and is easy to implemented for new
-/// types.
-///
-/// A variety of `From` impls are provided so that common pipe types are
-/// easy to create. For example:
-///
-/// ```
-/// use wasmtime_wasi::preview2::{pipe::ReadPipe, WasiCtxBuilder};
-/// let stdin = ReadPipe::from("hello from stdin!");
-/// let builder = WasiCtxBuilder::new().set_stdin(stdin);
-/// ```
+use bytes::Bytes;
+
 #[derive(Debug)]
-pub struct ReadPipe<R: Read + ReadReady> {
-    reader: Arc<RwLock<R>>,
+pub struct MemoryInputPipe {
+    buffer: std::io::Cursor<Bytes>,
 }
 
-impl<R: Read + ReadReady> Clone for ReadPipe<R> {
-    fn clone(&self) -> Self {
+impl MemoryInputPipe {
+    pub fn new(bytes: Bytes) -> Self {
         Self {
-            reader: self.reader.clone(),
+            buffer: std::io::Cursor::new(bytes),
         }
     }
-}
-
-impl<R: Read + ReadReady> ReadPipe<R> {
-    /// Create a new pipe from a `Read` type.
-    ///
-    /// All `Handle` read operations delegate to reading from this underlying reader.
-    pub fn new(r: R) -> Self {
-        Self::from_shared(Arc::new(RwLock::new(r)))
-    }
 
-    /// Create a new pipe from a shareable `Read` type.
-    ///
-    /// All `Handle` read operations delegate to reading from this underlying reader.
-    pub fn from_shared(reader: Arc<RwLock<R>>) -> Self {
-        Self { reader }
+    pub fn is_empty(&self) -> bool {
+        self.buffer.get_ref().len() as u64 == self.buffer.position()
     }
+}
 
-    /// Try to convert this `ReadPipe<R>` back to the underlying `R` type.
-    ///
-    /// This will fail with `Err(self)` if multiple references to the underlying `R` exist.
-    pub fn try_into_inner(mut self) -> Result<R, Self> {
-        match Arc::try_unwrap(self.reader) {
-            Ok(rc) => Ok(RwLock::into_inner(rc).unwrap()),
-            Err(reader) => {
-                self.reader = reader;
-                Err(self)
-            }
+#[async_trait::async_trait]
+impl HostInputStream for MemoryInputPipe {
+    fn read(&mut self, size: usize) -> Result<(Bytes, StreamState), Error> {
+        if self.is_empty() {
+            return Ok((Bytes::new(), StreamState::Closed));
         }
+
+        let mut dest = bytes::BytesMut::zeroed(size);
+        let nbytes = std::io::Read::read(&mut self.buffer, dest.as_mut())?;
+        dest.truncate(nbytes);
+
+        let state = if self.is_empty() {
+            StreamState::Closed
+        } else {
+            StreamState::Open
+        };
+        Ok((dest.freeze(), state))
     }
-    fn borrow(&self) -> std::sync::RwLockWriteGuard<R> {
-        RwLock::write(&self.reader).unwrap()
+    async fn ready(&mut self) -> Result<(), Error> {
+        if !self.is_empty() {
+            Ok(())
+        } else {
+            futures::future::pending().await
```

Comment:
I think that this may always want to be `Ok(())` since `read` never blocks?

Reply:
Correct.